src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs

```diff
@@ -199,30 +199,55 @@ private void UpdateSize()
                 if (extraField.Size < sizeof(long))
                     return true;
 
-                long value64 = reader.ReadInt64();
+                // Advancing the stream (by reading from it) is possible only when:
+                // 1. There is an explicit ask to do that (valid files, corresponding boolean flag(s) set to true).
+                // 2. When the size indicates that all the information is available ("slightly invalid files").
+                bool readAllFields = extraField.Size >= sizeof(long) + sizeof(long) + sizeof(long) + sizeof(int);
+
                 if (readUncompressedSize)
-                    zip64Block._uncompressedSize = value64;
+                {
+                    zip64Block._uncompressedSize = reader.ReadInt64();
+                }
+                else if (readAllFields)
+                {
+                    _ = reader.ReadInt64();
+                }
 
                 if (ms.Position > extraField.Size - sizeof(long))
                     return true;
 
-                value64 = reader.ReadInt64();
                 if (readCompressedSize)
-                    zip64Block._compressedSize = value64;
+                {
+                    zip64Block._compressedSize = reader.ReadInt64();
+                }
+                else if (readAllFields)
+                {
+                    _ = reader.ReadInt64();
+                }
 
                 if (ms.Position > extraField.Size - sizeof(long))
                     return true;
 
-                value64 = reader.ReadInt64();
                 if (readLocalHeaderOffset)
-                    zip64Block._localHeaderOffset = value64;
+                {
+                    zip64Block._localHeaderOffset = reader.ReadInt64();
+                }
+                else if (readAllFields)
+                {
+                    _ = reader.ReadInt64();
+                }
 
                 if (ms.Position > extraField.Size - sizeof(int))
                     return true;
 
-                int value32 = reader.ReadInt32();
                 if (readStartDiskNumber)
-                    zip64Block._startDiskNumber = value32;
+                {
+                    zip64Block._startDiskNumber = reader.ReadInt32();
+                }
+                else if (readAllFields)
+                {
+                    _ = reader.ReadInt32();
+                }
 
                 // original values are unsigned, so implies value is too big to fit in signed integer
                 if (zip64Block._uncompressedSize < 0) throw new InvalidDataException(SR.FieldTooBigUncompressedSize);
```

Comment:
These validations only occur if extraField.Size >= 28 since you are now returning if there's less than that. 

Shouldn't we validate when `reader.ReadInt*` is called and/or if read of the field was requested?

Reply:
This is how it always was, I would prefer not to do it now as I intend to backport it to 7.0 and trying to make it as defensive/safe as possible.

Reply:
The reason I was asking is because it wasn't always like that. It was changed to how it is now on https://github.com/dotnet/runtime/pull/68106, which is the PR that introduced the regression.